google/cloud/aiplatform/training_jobs.py

```diff
@@ -194,11 +217,68 @@ def get(
 
         return self
 
+    @classmethod
+    def _get_and_return_subclass(
+        cls,
+        resource_name: str,
+        project: Optional[str] = None,
+        location: Optional[str] = None,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ) -> "_TrainingJob":
+        """Retrieve Training Job subclass for the given resource_name without
+        knowing the training_task_definition.
+
+        Example usage:
+        ```
+        aiplatform.training_jobs._TrainingJob._get_and_return_subclass(
+            'projects/.../locations/.../trainingPipelines/12345'
+        )
+        # Returns: <google.cloud.aiplatform.training_jobs.AutoMLImageTrainingJob>
+        ```
+
+        Args:
+            resource_name (str):
+                Required. A fully-qualified resource name or ID.
+            project (str):
+                Optional project to retrieve dataset from. If not set, project
+                set in aiplatform.init will be used.
+            location (str):
+                Optional location to retrieve dataset from. If not set, location
+                set in aiplatform.init will be used.
+            credentials (auth_credentials.Credentials):
+                Optional. Custom credentials to use to upload this model. Overrides
+                credentials set in aiplatform.init.
+
+        Returns:
+            A Vertex AI Training Job
+        """
+
+        # Retrieve training pipeline resource before class construction
+        client = cls._instantiate_client(location=location, credentials=credentials)
+
+        gca_training_pipeline = getattr(client, cls._getter_method)(name=resource_name)
+
+        schema_uri = gca_training_pipeline.training_task_definition
+
+        # Collect all AutoML training job classes and CustomTrainingJob
+        class_list = [
```

Comment:
Prefer to not use a check against the classname as it may break after a rename.

Why wouldn't `class_list = cls.__subclasses__()` work?

Reply:
It's due to the fact we're depending on `_supported_training_schemas` in order to identify the correct subclass. This allows us to identify a given TrainingPipeline as either `AutoML[ObjectiveName]` or `Custom`.

Calling `cls.__subclasses__()` directly would cause hidden classes like `_CustomTrainingJob` to be returned, where as gathering all classes two levels deep would introduced things like `CustomPythonPackageTrainingJob`, etc. which we don't want to return.

The current logic collects the AutoML classes (which have designated `_supported_training_schemas`) and the primary `CustomTrainingJob` class. I could modify it to look for AutoML classes two levels deep.

Reply:
Since `_supported_training_schemas` isn't enough information to decide which subclass to use, I think you might as well just hardcode a dict to map `schema_uri` to `class`. This will be explicit.

The way you have now, if there are multiple classes that have the same schema in `_supported_training_schemas`, whatever is first in your list will be returned, which is less explicit (i.e. hard to debug).

Reply:
After discussing offline, I've added a test that ensures the uniqueness of `_supported_training_schemas` across all the potential subclasses. See [the commit](https://github.com/googleapis/python-aiplatform/pull/477/files/88c076a4334892f21529d77064e728cd6d74c0c3).